components/datetime/src/format.rs

```diff
@@ -0,0 +1,137 @@
+    let year = if month < 2 { year - 1 } else { year };
+    (year + year / 4 - year / 100 + year / 400 + t[month] + day + 1) % 7
+}
+
+pub fn write_pattern<T>(
+    pattern: &crate::pattern::Pattern,
+    data: &structs::dates::gregory::DatesV1,
+    date_time: &T,
+    w: &mut impl fmt::Write,
+) -> Result<(), DateTimeFormatError>
+where
+    T: DateTimeType,
+{
+    for item in &pattern.0 {
+        match item {
+            PatternItem::Field(field) => match field.symbol {
+                FieldSymbol::Year(..) => format_number(w, date_time.year(), &field.length)?,
+                FieldSymbol::Month(month) => match field.length {
+                    FieldLength::One | FieldLength::TwoDigit => {
+                        format_number(w, date_time.month() + 1, &field.length)?
+                    }
+                    length => {
+                        let symbol = data.get_symbol_for_month(month, length, date_time.month());
+                        w.write_str(symbol)?
+                    }
+                },
+                FieldSymbol::Weekday(weekday) => {
+                    let dow = get_day_of_week(date_time.year(), date_time.month(), date_time.day());
+                    let symbol = data.get_symbol_for_weekday(weekday, field.length, dow);
+                    w.write_str(symbol)?
+                }
+                FieldSymbol::Day(..) => format_number(w, date_time.day() + 1, &field.length)?,
+                FieldSymbol::Hour(hour) => {
+                    let value = match hour {
+                        fields::Hour::H11 => date_time.hour() % 12,
+                        fields::Hour::H12 => {
+                            let v = date_time.hour() % 12;
+                            if v == 0 {
+                                12
+                            } else {
+                                v
+                            }
+                        }
+                        fields::Hour::H23 => date_time.hour(),
+                        fields::Hour::H24 => {
+                            if date_time.hour() == 0 {
+                                24
+                            } else {
+                                date_time.hour()
+                            }
+                        }
+                    };
+                    format_number(w, value, &field.length)?
+                }
+                FieldSymbol::Minute => format_number(w, date_time.minute(), &field.length)?,
+                FieldSymbol::Second(..) => format_number(w, date_time.second(), &field.length)?,
+                FieldSymbol::DayPeriod(period) => match period {
+                    fields::DayPeriod::AmPm => {
+                        let symbol =
+                            data.get_symbol_for_day_period(period, field.length, date_time.hour());
+                        w.write_str(symbol)?
+                    }
+                    _ => unimplemented!(),
+                },
+                b => unimplemented!("{:#?}", b),
```

Comment:
Here and above: I'd rather not have `unimplemented!` here, because this is a data-driven code path, not a user code path.  I think the only thing that triggers this code path is era.  Can we fail earlier (in the pattern parser) by not allowing patterns to be consumed that we can't format?

Reply:
yup, just removed Era for now. We can re-add it when we get to skeletons :)